#Problem Statement
Northern Lights Air launched a loyalty-program promotion from February through April 2018 to increase customer enrollment. The goal of this analysis is to evaluate enrollment during the promotion and determine whether customers who joined during this period demonstrated different levels of retention, flight engagement, points activity, and customer lifetime value (CLV) compared with relevant historical customers.

The analysis also examines customer segments to identify patterns associated with higher customer value.

#Background
Northern Lights Air (NLA) is a fictional Canadian airline with a customer loyalty program that tracks enrollment, cancellations, flight activity, loyalty points, and customer demographics. From February through April 2018, NLA conducted a promotional campaign designed to increase loyalty-program enrollment.

#Business Objective
Northern Lights Air conducted a loyalty program promotion from February through April 2018. The objective of this analysis is to evaluate the effectiveness of the promotion and better understand the customers participating in the loyalty program.

The analysis will focus on answering the following business questions:
* Did loyalty-program enrollment increase during the promotion compared with previous years?
* How well were customers acquired during the promotion retained?
* How engaged were promotion customers through flight activity and loyalty-point activity?
* How does the value of promotion customers compare with a relevant historical customer group?
* Which customer segments are associated with higher customer lifetime value?

The goal is to provide NLA with insights that can help inform future loyalty promotions, customer retention efforts, and customer targeting strategies.

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving Calendar.csv to Calendar.csv
Saving Customer Flight Activity.csv to Customer Flight Activity.csv
Saving Customer Loyalty History.csv to Customer Loyalty History.csv


In [ ]:
import pandas as pd
import sqlite3

In [ ]:
loyalty = pd.read_csv("Customer Loyalty History.csv")
flights = pd.read_csv("Customer Flight Activity.csv")
calendar = pd.read_csv("Calendar.csv")

In [ ]:
print("Loyalty:", loyalty.shape)
print("Flights:", flights.shape)
print("Calendar:", calendar.shape)

Loyalty: (16737, 16)
Flights: (392936, 8)
Calendar: (2557, 4)


In [ ]:
for df in [loyalty, flights, calendar]:
    df.columns = (
        df.columns
        .str.strip()
        .str.lower()
        .str.replace(" ", "_")
    )

In [ ]:
loyalty.columns.tolist()

['loyalty_number',
 'country',
 'province',
 'city',
 'postal_code',
 'gender',
 'education',
 'salary',
 'marital_status',
 'loyalty_card',
 'clv',
 'enrollment_type',
 'enrollment_year',
 'enrollment_month',
 'cancellation_year',
 'cancellation_month']

In [ ]:
flights.columns.tolist()

['loyalty_number',
 'year',
 'month',
 'total_flights',
 'distance',
 'points_accumulated',
 'points_redeemed',
 'dollar_cost_points_redeemed']

In [ ]:
calendar.columns.tolist()

['date', 'start_of_year', 'start_of_quarter', 'start_of_month']

#SQL

##Introduction

In [ ]:
conn = sqlite3.connect("airline_loyalty.db")

loyalty.to_sql(
    "loyalty_history",
    conn,
    if_exists="replace",
    index=False
)

flights.to_sql(
    "flight_activity",
    conn,
    if_exists="replace",
    index=False
)

calendar.to_sql(
    "calendar",
    conn,
    if_exists="replace",
    index=False
)

2557

In [ ]:
query = """
SELECT *
FROM loyalty_history
"""
pd.read_sql_query(query,conn)

,loyalty_number,country,province,city,postal_code,gender,education,salary,marital_status,loyalty_card,clv,enrollment_type,enrollment_year,enrollment_month,cancellation_year,cancellation_month
0,480934,Canada,Ontario,Toronto,M2Z 4K1,Female,Bachelor,83236.0,Married,Star,3839.14,Standard,2016,2,NaN,NaN
1,549612,Canada,Alberta,Edmonton,T3G 6Y6,Male,College,NaN,Divorced,Star,3839.61,Standard,2016,3,NaN,NaN
2,429460,Canada,British Columbia,Vancouver,V6E 3D9,Male,College,NaN,Single,Star,3839.75,Standard,2014,7,2018.0,1.0
3,608370,Canada,Ontario,Toronto,P1W 1K4,Male,College,NaN,Single,Star,3839.75,Standard,2013,2,NaN,NaN
4,530508,Canada,Quebec,Hull,J8Y 3Z5,Male,Bachelor,103495.0,Married,Star,3842.79,Standard,2014,10,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
16732,823768,Canada,British Columbia,Vancouver,V6E 3Z3,Female,College,NaN,Married,Star,61850.19,Standard,2012,12,NaN,NaN
16733,680886,Canada,Saskatchewan,Regina,S1J 3C5,Female,Bachelor,89210.0,Married,Star,67907.27,Standard,2014,9,NaN,NaN
16734,776187,Canada,British Columbia,Vancouver,V5R 1W3,Male,College,NaN,Single,Star,74228.52,Standard,2014,3,NaN,NaN
16735,906428,Canada,Yukon,Whitehorse,Y2K 6R0,Male,Bachelor,-57297.0,Married,Star,10018.66,2018 Promotion,2018,4,NaN,NaN


In [ ]:
query = """
SELECT *
FROM flight_activity
"""
pd.read_sql_query(query,conn)

,loyalty_number,year,month,total_flights,distance,points_accumulated,points_redeemed,dollar_cost_points_redeemed
0,100590,2018,6,12,15276,22914.0,0,0
1,100590,2018,7,12,9168,13752.0,0,0
2,100590,2018,5,4,6504,9756.0,0,0
3,100590,2018,10,0,0,0.0,512,92
4,100590,2018,2,0,0,0.0,0,0
...,...,...,...,...,...,...,...,...
392931,999986,2018,4,0,0,0.0,0,0
392932,999986,2018,5,0,0,0.0,0,0
392933,999986,2018,6,0,0,0.0,0,0
392934,999986,2018,9,0,0,0.0,0,0


In [ ]:
query = """
SELECT *
FROM calendar
"""
pd.read_sql_query(query,conn)

,date,start_of_year,start_of_quarter,start_of_month
0,2012-01-01,2012-01-01,2012-01-01,2012-01-01
1,2012-01-02,2012-01-01,2012-01-01,2012-01-01
2,2012-01-03,2012-01-01,2012-01-01,2012-01-01
3,2012-01-04,2012-01-01,2012-01-01,2012-01-01
4,2012-01-05,2012-01-01,2012-01-01,2012-01-01
...,...,...,...,...
2552,2018-12-27,2018-01-01,2018-10-01,2018-12-01
2553,2018-12-28,2018-01-01,2018-10-01,2018-12-01
2554,2018-12-29,2018-01-01,2018-10-01,2018-12-01
2555,2018-12-30,2018-01-01,2018-10-01,2018-12-01


In [ ]:
query = """
SELECT COUNT(*) AS loyalty_rows
FROM loyalty_history
"""
pd.read_sql_query(query,conn)

,loyalty_rows
0,16737


In [ ]:
query = """
SELECT COUNT(DISTINCT loyalty_number) AS unique_loyalty_count
FROM loyalty_history
"""
pd.read_sql_query(query,conn)


,unique_loyalty_count
0,16737


In [ ]:
query = """
SELECT COUNT(*) AS flight_rows
FROM flight_activity
"""
pd.read_sql_query(query,conn)

,flight_rows
0,392936


##Unique Values

In [ ]:
query = """
SELECT COUNT(DISTINCT loyalty_number) AS unique_loyalty_count
FROM flight_activity
"""
pd.read_sql_query(query,conn)

,unique_loyalty_count
0,16737


In [ ]:
query = """
SELECT COUNT(*) AS unique_flight_count
FROM (
  SELECT DISTINCT loyalty_number, month, year
  FROM flight_activity) AS unique_flight_activity
"""
pd.read_sql_query(query,conn)

,unique_flight_count
0,389065


In [ ]:
query = """
SELECT
  loyalty_number,
  month,
  year,
  COUNT (*) AS number_of_records
FROM flight_activity
GROUP BY loyalty_number, month, year
HAVING number_of_records > 1
"""
pd.read_sql_query(query,conn)

,loyalty_number,month,year,number_of_records
0,101902,1,2017,2
1,101902,1,2018,2
2,101902,2,2017,2
3,101902,2,2018,2
4,101902,3,2017,2
...,...,...,...,...
3842,992168,10,2018,2
3843,992168,11,2017,2
3844,992168,11,2018,2
3845,992168,12,2017,2


In [ ]:
query = """
SELECT *
FROM flight_activity
WHERE loyalty_number = 101902 AND month = 1 AND year = 2017
"""
pd.read_sql_query(query,conn)

,loyalty_number,year,month,total_flights,distance,points_accumulated,points_redeemed,dollar_cost_points_redeemed
0,101902,2017,1,0,0,0.0,0,0
1,101902,2017,1,0,0,0.0,0,0


##Duplicates

In [ ]:
query = """
SELECT COUNT(*) AS duplicate_flight_count
FROM (
  SELECT DISTINCT *
  FROM flight_activity)
"""
pd.read_sql_query(query,conn)

,duplicate_flight_count
0,391014


In [ ]:
query = """
SELECT
  loyalty_number,
  month,
  year,
  COUNT (DISTINCT total_flights) AS distinct_flight_values
FROM flight_activity
GROUP BY loyalty_number, month, year
HAVING distinct_flight_values > 1
"""
pd.read_sql_query(query,conn)

,loyalty_number,month,year,distinct_flight_values
0,101902,1,2018,2
1,101902,2,2018,2
2,101902,3,2018,2
3,101902,4,2018,2
4,101902,5,2017,2
...,...,...,...,...
1748,992168,10,2017,2
1749,992168,11,2017,2
1750,992168,11,2018,2
1751,992168,12,2017,2


In [ ]:
query = """
SELECT *
FROM flight_activity
WHERE loyalty_number = 992168 AND month = 10 AND year = 2017
"""
pd.read_sql_query(query,conn)

,loyalty_number,year,month,total_flights,distance,points_accumulated,points_redeemed,dollar_cost_points_redeemed
0,992168,2017,10,3,2205,2205.0,0,0
1,992168,2017,10,1,2398,2398.0,0,0


##Missing Values

In [ ]:
query = """
SELECT COUNT(*) AS missing_salary
FROM loyalty_history
WHERE salary IS NULL
"""
pd.read_sql_query(query,conn)


,missing_salary
0,4238


In [ ]:
query = """
SELECT education, COUNT(*) AS missing_salary
FROM loyalty_history
WHERE salary IS NULL
GROUP BY education
"""
pd.read_sql_query(query,conn)

,education,missing_salary
0,College,4238


In [ ]:
query = """
SELECT COUNT(*) AS missing_cancellation_date
FROM loyalty_history
WHERE cancellation_year IS NULL AND cancellation_month IS NULL
"""
pd.read_sql_query(query,conn)

,missing_cancellation_date
0,14670


###Did loyalty-program enrollment increase during the February–April 2018 promotion?

In [ ]:
query = """
SELECT
  enrollment_month,
  enrollment_year,
  COUNT(*) AS number_of_enrollments
FROM loyalty_history
GROUP BY enrollment_month,enrollment_year
ORDER BY enrollment_year, enrollment_month
"""
pd.read_sql_query(query,conn)

,enrollment_month,enrollment_year,number_of_enrollments
0,4,2012,62
1,5,2012,203
2,6,2012,195
3,7,2012,213
4,8,2012,206
...,...,...,...
76,8,2018,218
77,9,2018,185
78,10,2018,213
79,11,2018,246


In [ ]:
query = """
SELECT
  enrollment_month,
  enrollment_year,
  COUNT(*) AS number_of_enrollments
FROM loyalty_history
WHERE enrollment_month IN (2,3,4)
GROUP BY enrollment_month,enrollment_year
ORDER BY enrollment_year, enrollment_month
"""
pd.read_sql_query(query,conn)

,enrollment_month,enrollment_year,number_of_enrollments
0,4,2012,62
1,2,2013,175
2,3,2013,201
3,4,2013,192
4,2,2014,189
5,3,2014,199
6,4,2014,178
7,2,2015,184
8,3,2015,183
9,4,2015,200


###How did enrollment during Feb–Apr 2018 compare with the same months in previous years?

In [ ]:
query = """
WITH monthly_enrollment_counts AS (
  SELECT
    enrollment_month,
    enrollment_year,
    COUNT(*) AS number_of_enrollments
  FROM loyalty_history
  WHERE enrollment_month IN (2,3,4)
  GROUP BY enrollment_month, enrollment_year),

historical_average AS (
  SELECT
    enrollment_month,
    AVG(number_of_enrollments) AS avg_enrollments
  FROM monthly_enrollment_counts
  WHERE enrollment_year BETWEEN 2013 AND 2017
  GROUP BY enrollment_month)

SELECT
  historical_average.enrollment_month,
  avg_enrollments,
  monthly_enrollment_counts.number_of_enrollments AS promotion_enrollments
FROM historical_average
LEFT JOIN monthly_enrollment_counts
  ON historical_average.enrollment_month = monthly_enrollment_counts.enrollment_month
WHERE monthly_enrollment_counts.enrollment_year == 2018
"""
pd.read_sql_query(query, conn)

,enrollment_month,avg_enrollments,promotion_enrollments
0,2,185.0,295
1,3,205.6,330
2,4,196.0,346


###By what percentage did enrollment increase during each promotion month compared with its historical average?

In [ ]:
query = """
WITH monthly_enrollment_counts AS (
  SELECT
    enrollment_month,
    enrollment_year,
    COUNT(*) AS number_of_enrollments
  FROM loyalty_history
  WHERE enrollment_month IN (2,3,4)
  GROUP BY enrollment_month, enrollment_year),

historical_average AS (
  SELECT
    enrollment_month,
    AVG(number_of_enrollments) AS avg_enrollments
  FROM monthly_enrollment_counts
  WHERE enrollment_year BETWEEN 2013 AND 2017
  GROUP BY enrollment_month)

SELECT
  historical_average.enrollment_month,
  avg_enrollments,
  monthly_enrollment_counts.number_of_enrollments AS promotion_enrollments,
  ROUND(((monthly_enrollment_counts.number_of_enrollments - avg_enrollments)/avg_enrollments) * 100, 1) AS percent_increase
FROM historical_average
LEFT JOIN monthly_enrollment_counts
  ON historical_average.enrollment_month = monthly_enrollment_counts.enrollment_month
WHERE monthly_enrollment_counts.enrollment_year == 2018
"""
pd.read_sql_query(query, conn)

,enrollment_month,avg_enrollments,promotion_enrollments,percent_increase
0,2,185.0,295,59.5
1,3,205.6,330,60.5
2,4,196.0,346,76.5


###How many customers in each enrollment type have a recorded cancellation?

In [ ]:
query = """
SELECT
  enrollment_type,
  COUNT(*) AS total_customers,
  COUNT(cancellation_year) AS cancelled_customers
FROM
  loyalty_history
GROUP BY enrollment_type
"""
pd.read_sql_query(query,conn)

,enrollment_type,total_customers,cancelled_customers
0,2018 Promotion,971,115
1,Standard,15766,1952


###What percentage of customers in each enrollment type cancelled?

In [ ]:
query = """
SELECT
  enrollment_type,
  COUNT(*) AS total_customers,
  COUNT(cancellation_year) AS cancelled_customers,
  ROUND((COUNT(cancellation_year) * 100.0/COUNT(*)), 1) AS cancellation_rate
FROM
  loyalty_history
GROUP BY enrollment_type
"""
pd.read_sql_query(query,conn)

,enrollment_type,total_customers,cancelled_customers,cancellation_rate
0,2018 Promotion,971,115,11.8
1,Standard,15766,1952,12.4


###Of the 971 customers acquired through the promotion, how many remained enrolled through the end of the dataset?

In [ ]:
query = """
SELECT
  enrollment_type,
  COUNT(*) AS total_customers,
  COUNT(cancellation_year) AS cancelled_customers,
  COUNT(*) - COUNT(cancellation_year) AS customers_retained,
  ROUND(((COUNT(*) - COUNT(cancellation_year))* 100.0/COUNT(*)), 1) AS observed_retention_rate
FROM
  loyalty_history
WHERE
  enrollment_type = '2018 Promotion'
GROUP BY enrollment_type
"""
pd.read_sql_query(query,conn)

,enrollment_type,total_customers,cancelled_customers,customers_retained,observed_retention_rate
0,2018 Promotion,971,115,856,88.2


###Did customers acquired through the 2018 Promotion fly more or less than other customers?

In [ ]:
query = """
SELECT
  loyalty_history.enrollment_type,
  COUNT(DISTINCT loyalty_history.loyalty_number) AS total_unique_customers,
  SUM(flight_activity.total_flights) AS flight_total,
  ROUND(SUM(flight_activity.total_flights) * 1.0/COUNT(DISTINCT loyalty_history.loyalty_number), 1) AS avg_flights_per_customer
FROM loyalty_history
LEFT JOIN flight_activity
  ON loyalty_history.loyalty_number = flight_activity.loyalty_number
WHERE
  loyalty_history.enrollment_type = '2018 Promotion' AND flight_activity.year = 2018
  AND flight_activity.month >= loyalty_history.enrollment_month
GROUP BY loyalty_history.enrollment_type
"""
pd.read_sql_query(query,conn)

,enrollment_type,total_unique_customers,flight_total,avg_flights_per_customer
0,2018 Promotion,971,44402,45.7


###How did the flight activity of the 2018 Promotion cohort compare with customers who enrolled during Feb–Apr in previous years?

In [ ]:
query = """
SELECT
  enrollment_year,
  COUNT(DISTINCT loyalty_history.loyalty_number) AS total_unique_customers,
  SUM(flight_activity.total_flights) AS flight_total,
  ROUND(SUM(flight_activity.total_flights) * 1.0/COUNT(DISTINCT loyalty_history.loyalty_number), 1) AS avg_flights_per_customer
FROM
  loyalty_history
LEFT JOIN flight_activity
  ON loyalty_history.loyalty_number = flight_activity.loyalty_number
WHERE
  enrollment_type = '2018 Promotion'
  AND (
    (flight_activity.year - loyalty_history.enrollment_year) * 12
    + (flight_activity.month - loyalty_history.enrollment_month)
) BETWEEN 0 AND 8
GROUP BY enrollment_year
"""
pd.read_sql_query (query,conn)

,enrollment_year,total_unique_customers,flight_total,avg_flights_per_customer
0,2018,971,40549,41.8


In [ ]:
query = """
WITH clean_flight_activity AS (
  SELECT DISTINCT *
  FROM
    flight_activity
  ),

cohort_analysis AS (
  SELECT
    enrollment_year,
    COUNT(DISTINCT loyalty_history.loyalty_number) AS total_unique_customers,
    SUM(clean_flight_activity.total_flights) AS flight_total,
    ROUND(SUM(clean_flight_activity.total_flights) * 1.0/COUNT(DISTINCT loyalty_history.loyalty_number), 1) AS avg_flights_per_customer
  FROM
    loyalty_history
  LEFT JOIN clean_flight_activity
    ON loyalty_history.loyalty_number = clean_flight_activity.loyalty_number
  WHERE
    (
      (loyalty_history.enrollment_year = 2017 AND loyalty_history.enrollment_month IN (2,3,4)) OR loyalty_history.enrollment_type = '2018 Promotion'
      )
    AND (
      (clean_flight_activity.year - loyalty_history.enrollment_year) * 12
      + (clean_flight_activity.month - loyalty_history.enrollment_month)
  ) BETWEEN 0 AND 8
  GROUP BY enrollment_year)
SELECT
 (SELECT avg_flights_per_customer
 FROM cohort_analysis
 WHERE enrollment_year = 2018) AS twenty_eighteen_average,

 (SELECT avg_flights_per_customer
  FROM cohort_analysis
  WHERE enrollment_year = 2017) AS twenty_seventeen_average,

 Round((((SELECT avg_flights_per_customer
 FROM cohort_analysis
 WHERE enrollment_year = 2018)) - (SELECT avg_flights_per_customer
  FROM cohort_analysis
  WHERE enrollment_year = 2017)) * 100.0/(SELECT avg_flights_per_customer
  FROM cohort_analysis
  WHERE enrollment_year = 2017), 1) AS cohort_comparison
FROM
  cohort_analysis

"""
pd.read_sql_query(query,conn)

,twenty_eighteen_average,twenty_seventeen_average,cohort_comparison
0,41.8,13.9,200.7
1,41.8,13.9,200.7


###Did customers acquired through the 2018 Promotion have different average CLV than a comparable prior-year cohort?

In [ ]:
query = """
SELECT
  enrollment_year,
  COUNT(DISTINCT loyalty_number) AS total_unique_customers,
  ROUND(AVG(clv),2) AS avg_clv
FROM
  loyalty_history
WHERE
  (enrollment_year = 2017 AND enrollment_month IN (2,3,4)) OR enrollment_type = '2018 Promotion'
GROUP BY
  enrollment_year
"""
pd.read_sql_query(query,conn)

,enrollment_year,total_unique_customers,avg_clv
0,2017,629,7434.70
1,2018,971,8046.51


In [ ]:
query = """
WITH clv_analysis AS (
  SELECT
  enrollment_year,
  COUNT(DISTINCT loyalty_number) AS total_unique_customers,
  ROUND(AVG(clv),2) AS avg_clv
FROM
  loyalty_history
WHERE
  (enrollment_year = 2017 AND enrollment_month IN (2,3,4)) OR enrollment_type = '2018 Promotion'
GROUP BY
  enrollment_year)

SELECT
  (SELECT avg_clv
  FROM clv_analysis
  WHERE enrollment_year = 2018) AS clv_twenty_eighteen_avg,

  (SELECT avg_clv
  FROM clv_analysis
  WHERE enrollment_year = 2017) AS clv_twenty_seventeen_avg,

  ROUND (((SELECT avg_clv
  FROM clv_analysis
  WHERE enrollment_year = 2018) - (SELECT avg_clv
  FROM clv_analysis
  WHERE enrollment_year = 2017))*100.0/(SELECT avg_clv
  FROM clv_analysis
  WHERE enrollment_year = 2017),1) AS percent_increase

"""

pd.read_sql_query(query,conn)

,clv_twenty_eighteen_avg,clv_twenty_seventeen_avg,percent_increase
0,8046.51,7434.7,8.2


###Did the 2018 promotion cohort show greater loyalty-program activity than the comparable 2017 cohort?

In [ ]:
query = """
WITH clean_flight_activity AS (
    SELECT DISTINCT *
    FROM flight_activity
),

points_analysis AS (
  SELECT
    loyalty_history.enrollment_year,
    COUNT(DISTINCT loyalty_history.loyalty_number) AS total_unique_customers,
    SUM(clean_flight_activity.points_accumulated) AS total_points_accumulated,
    ROUND(SUM(clean_flight_activity.points_accumulated) * 1.0/COUNT(DISTINCT loyalty_history.loyalty_number)) AS avg_points_per_customer,
    ROUND(SUM(clean_flight_activity.points_redeemed) * 1.0/COUNT(DISTINCT loyalty_history.loyalty_number)) AS avg_points_redeemed_per_customer
  FROM loyalty_history
  LEFT JOIN clean_flight_activity
    ON loyalty_history.loyalty_number = clean_flight_activity.loyalty_number

  WHERE
    (
      (loyalty_history.enrollment_year = 2017 AND loyalty_history.enrollment_month IN (2,3,4)) OR loyalty_history.enrollment_type = '2018 Promotion'
      )
    AND (
      (clean_flight_activity.year - loyalty_history.enrollment_year) * 12
      + (clean_flight_activity.month - loyalty_history.enrollment_month)
  ) BETWEEN 0 AND 8

  GROUP BY enrollment_year)

SELECT *
FROM points_analysis

"""

pd.read_sql_query(query,conn)






,enrollment_year,total_unique_customers,total_points_accumulated,avg_points_per_customer,avg_points_redeemed_per_customer
0,2017,629,13071700.0,20782.0,327.0
1,2018,971,91063362.0,93783.0,333.0


In [ ]:
query = """
WITH clean_flight_activity AS (
    SELECT DISTINCT *
    FROM flight_activity
),

points_analysis AS (
  SELECT
    loyalty_history.enrollment_year,
    COUNT(DISTINCT loyalty_history.loyalty_number) AS total_unique_customers,
    SUM(clean_flight_activity.points_accumulated) AS total_points_accumulated,
    ROUND(SUM(clean_flight_activity.points_accumulated) * 1.0/COUNT(DISTINCT loyalty_history.loyalty_number)) AS avg_points_per_customer,
    ROUND(SUM(clean_flight_activity.points_redeemed) * 1.0/COUNT(DISTINCT loyalty_history.loyalty_number)) AS avg_points_redeemed_per_customer
  FROM loyalty_history
  LEFT JOIN clean_flight_activity
    ON loyalty_history.loyalty_number = clean_flight_activity.loyalty_number

  WHERE
    (
      (loyalty_history.enrollment_year = 2017 AND loyalty_history.enrollment_month IN (2,3,4)) OR loyalty_history.enrollment_type = '2018 Promotion'
      )
    AND (
      (clean_flight_activity.year - loyalty_history.enrollment_year) * 12
      + (clean_flight_activity.month - loyalty_history.enrollment_month)
  ) BETWEEN 0 AND 8

  GROUP BY enrollment_year)

SELECT
  MAX(CASE WHEN enrollment_year = 2018
         THEN avg_points_per_customer
    END) AS twenty_eighteen_avg_points,
  MAX(CASE WHEN enrollment_year = 2017
         THEN avg_points_per_customer
    END) AS twenty_seventeen_avg_points,
  ROUND((MAX(CASE WHEN enrollment_year = 2018
         THEN avg_points_per_customer
    END) - MAX(CASE WHEN enrollment_year = 2017
         THEN avg_points_per_customer
    END)) * 100.0/MAX(CASE WHEN enrollment_year = 2017
         THEN avg_points_per_customer
    END),1) AS percent_increase
FROM points_analysis

"""

pd.read_sql_query(query,conn)

,twenty_eighteen_avg_points,twenty_seventeen_avg_points,percent_increase
0,93783.0,20782.0,351.3


#Download Files

In [ ]:
clean_flight_activity = pd.read_sql_query(
    """
    SELECT DISTINCT *
    FROM flight_activity
    """,
    conn
)

In [ ]:
clean_flight_activity.shape

(391014, 8)

In [ ]:
loyalty_history = pd.read_sql_query(
    """
    SELECT *
    FROM loyalty_history
    """,
    conn
)

In [ ]:
calendar = pd.read_sql_query(
    """
    SELECT *
    FROM calendar
    """,
    conn
)

In [ ]:
loyalty_history.shape

(16737, 16)

In [ ]:
calendar.shape

(2557, 4)

In [ ]:
loyalty_history.to_csv("loyalty_history.csv", index=False)

clean_flight_activity.to_csv("clean_flight_activity.csv", index=False)

calendar.to_csv("calendar.csv", index=False)

In [ ]:
from google.colab import files

files.download("loyalty_history.csv")
files.download("clean_flight_activity.csv")
files.download("calendar.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>